### DISCUSSION TITLE
/alerts/active/count v.s. /alerts/active?status=actual for zones inconsistency

### DISCUSSION THREAD

**Discussion by ktrue** [weather-gov/api#665] [Show and tell]
I've written a script that displays current alerts on GRLevel3 radar software https://github.com/ktrue/NWS-alerts-placefile

Sometimes, /alerts/active?status=actual has alerts with UGC entries for zones involved, but seems to occasionally miss including zones (particularly for Marine alerts) in the UGC entries.  The map at www.weather.gov and the www.weather.gov/gis/ show the shading for alerts correctly (but are not using api.weather.gov for data.. they're using the mapserver data instead).

I wrote a small PHP script to fetch the counts, land and marine alerts set (divided by land and marine to stay below the limit=500 on the queries) and noticed the following:
1) the counts of alerts do not match the number of alerts returned by the active queries
2) zones listed in the counts do not all appear in the alerts UGC entries
3) alerts do not have all the UgC zones listed in the counts
4) for zones appearing in both counts and alerts, the counts of appearances differs for many zones listed

I've attached the script .. run with check-api.php?force to fetch the current data, omit the ?force to use the cached data.

sample output (snipped):
```
check-api.php - V1.01 - 12-Oct-2023
 fetching https://api.weather.gov/alerts/active/count
 saved 27113 bytes to alert-counts.json
 JSON decode - No errors

 fetching https://api.weather.gov/alerts/active?status=actual&message_type=alert&region_type=land&limit=500
 saved 903788 bytes to alert-land.json
 JSON decode - No errors

 fetching https://api.weather.gov/alerts/active?status=actual&message_type=alert&region_type=marine&limit=500
 saved 1030363 bytes to alert-marine.json
 JSON decode - No errors

 using current UTC as time
 now UTC for comparisons is Thu, 12 Oct 2023 19:58:52 +0000
 alerts stats will include expired ones.
 alerts/active/count   land=514 v.s. count=225 alert features
 alerts/active/count marine=339 v.s. count=335 alert features
 alerts/active/count  zones=1248 with alerts
 alerts/active/count has 2104 total alerts from 1248 zones listed.

 Analyzing LAND alerts
0: Special Weather Statement issued October 12 at 2:58PM CDT by NWS Hastings NE
    severity=Moderate, certainty=Observed, urgency=Expected
    expires=Thu, 12 Oct 2023 20:30:00 +0000
    ends=   Thu, 12 Oct 2023 20:30:00 +0000 expired=no
    polygon=YES zones=KSZ005, NEZ082, NEZ083
1: Flash Flood Warning issued October 12 at 2:57PM CDT until October 12 at 9:00PM CDT by NWS North Platte NE
    severity=Severe, certainty=Likely, urgency=Immediate
    expires=Fri, 13 Oct 2023 02:00:00 +0000
    ends=   Fri, 13 Oct 2023 02:00:00 +0000 expired=no
    polygon=YES zones=NEC009, NEC041, NEC115
...
223: Rip Current Statement issued October 9 at 4:43AM AST until October 12 at 6:00PM AST by NWS San Juan PR
    severity=Moderate, certainty=Likely, urgency=Expected
    expires=Mon, 09 Oct 2023 16:45:00 +0000
    ends=   Thu, 12 Oct 2023 22:00:00 +0000 expired=no
    polygon=no zones=PRZ001, PRZ002, PRZ005, PRZ008, PRZ012, VIZ002
224: High Wind Watch issued October 8 at 12:30PM AKDT until October 12 at 5:00PM AKDT by NWS Anchorage AK
    severity=Severe, certainty=Possible, urgency=Future
    expires=Tue, 10 Oct 2023 00:00:00 +0000
    ends=   Fri, 13 Oct 2023 01:00:00 +0000 expired=no
    polygon=no zones=AKZ187
..Done - 549 seenZones count

 Analyzing MARINE alerts
0: Gale Warning issued October 12 at 2:28PM CDT until October 14 at 8:00PM CDT by NWS Milwaukee/Sullivan WI
    severity=Moderate, certainty=Likely, urgency=Expected
    expires=Fri, 13 Oct 2023 03:30:00 +0000
    ends=   Sun, 15 Oct 2023 01:00:00 +0000 expired=no
    polygon=no zones=LMZ080, LMZ366, LMZ565, LMZ567, LMZ669, LMZ671, LMZ673, LMZ675, LMZ777,
    LMZ779, LMZ868, LMZ870, LMZ872, LMZ874, LMZ876, LMZ878
1: Gale Watch issued October 12 at 2:28PM CDT until October 14 at 7:00PM CDT by NWS Milwaukee/Sullivan WI
    severity=Moderate, certainty=Possible, urgency=Future
    expires=Fri, 13 Oct 2023 03:30:00 +0000
    ends=   Sun, 15 Oct 2023 00:00:00 +0000 expired=no
    polygon=no zones=LMZ645, LMZ646
...
333: Heavy Freezing Spray Warning issued October 6 at 3:28PM AKDT until October 14 at 5:00AM AKDT by NWS Fairbanks AK
    severity=Severe, certainty=Likely, urgency=Expected
    expires=Sat, 07 Oct 2023 14:30:00 +0000
    ends=   Sat, 14 Oct 2023 13:00:00 +0000 expired=no
    polygon=no zones=PKZ815
334: Heavy Freezing Spray Warning issued October 6 at 2:00AM AKDT until October 13 at 5:00PM AKDT by NWS Fairbanks AK
    severity=Severe, certainty=Likely, urgency=Expected
    expires=Sat, 07 Oct 2023 02:15:00 +0000
    ends=   Sat, 14 Oct 2023 01:00:00 +0000 expired=no
    polygon=no zones=PKZ815
..Done - 740 seenZones count total

 Checking Zone counts in /alerts/active/count v.s. zones seen in Land and Marine alerts UGC entries.

 These zones had #alerts found in /alerts/active/count but different #alerts counted in seenZones from UGC entries in alerts.
    AKZ187:2<3, AKZ323:2<3, AKZ327:1<2, AKZ328:1<2, AMZ350:3<5, AMZ352:3<5,
    AMZ354:3<5, AMZ374:3<5, AMZ472:2>1, AMZ474:2>1, COZ001:1<2, COZ008:1<2,
    COZ020:1<2, COZ021:1<2, COZ022:1<2, COZ023:1<2, COZ046:2<3, COZ047:2<3,
...
    WYZ009:1<2, WYZ015:1<2, WYZ022:1<2, WYZ057:1<2, WYZ103:1<2, WYZ104:1<2,
    WYZ105:1<2, WYZ112:1<2, WYZ114:1<2, WYZ116:1<2, WYZ198:1<2

 The following zones were in/alerts/active/count but not in seenZones from UGC entries in alerts.
    AKC016, AKC100, AKC105, AKC110, AKC198, AKC220, AKC230, AKC282, ALC003,
    ALC097, AMZ450, AMZ470, AMZ550, AMZ552, AMZ555, AMZ570, AMZ572, AMZ650,
    AMZ651, AMZ670, AMZ671, AZC001, AZC005, AZC017, CAC059, CAC065, CAC071,
...
    WYC013, WYC017, WYC019, WYC021, WYC023, WYC025, WYC029, WYC031, WYC033,
    WYC035, WYC037, WYC039, WYC043, WYC045
..Done. 787 mismatches found.

 Checking Zones seen in Land and Marine alerts UGC entries v.s /alerts/active/count zones.

 These zones had #alerts counted in seenZones from UGC entries in alerts but different #alerts found in /alerts/active/count .
    AKZ187:3>2, AKZ323:3>2, AKZ327:2>1, AKZ328:2>1, AMZ350:5>3, AMZ352:5>3,
    AMZ354:5>3, AMZ374:5>3, AMZ472:1<2, AMZ474:1<2, COZ001:2>1, COZ008:2>1,
    COZ020:2>1, COZ021:2>1, COZ022:2>1, COZ023:2>1, COZ046:3>2, COZ047:3>2,
...
    WYZ009:2>1, WYZ015:2>1, WYZ022:2>1, WYZ057:2>1, WYZ103:2>1, WYZ104:2>1,
    WYZ105:2>1, WYZ112:2>1, WYZ114:2>1, WYZ116:2>1, WYZ198:2>1

  The following zones were in seenZone but not in /alerts/active/count
    AMZ330, COZ234, COZ254, MDZ017, OKZ016, OKZ021, OKZ022, PZZ650, SDZ025,
    SDZ029
..Done. 279 mismatches found.
```

[check-api.zip](https://github.com/weather-gov/api/files/12886674/check-api.zip)

**Comment by sullynole** [comment]
/alerts/active?status=actual&message_type=alert&region_type=land&limit=500

Not sure I completely follow the comparisons being made here, but I would not expect this query to match the count endpoint numbers since it is filtering out messageType and status.  Also, the limit currently is not recognized on the active endpoint so it is not needed.

https://vlab.noaa.gov/web/nws-common-alerting-protocol/cap-documentation#msgtype
https://vlab.noaa.gov/web/nws-common-alerting-protocol/cap-documentation#status

**Reply by sullynole** [reply]
Note, the API active endpoint is the upstream source for the maps at http://www.weather.gov/gis/ (https://mapservices.weather.noaa.gov/eventdriven/rest/services/WWA/watch_warn_adv/MapServer) and radar.weather.gov .

**Comment by ktrue** [comment]
Since the documentation for /alerts/active/count says
```
A data structure showing the counts of active alerts broken down by various categories
```
the filters on /alerts/active?status=actual&message_type=alert should match (unless there are alerts counted that are message_type=alert with a status other than actual).

When doing the query /alerts/active?type=land and ?type=marine, shouldn't the counts for those in /alerts/active/count match the number of events in the detailed queries?

Then there's the whole thing about zones mentioned in the /alerts/active/count that do not appear in the UGC parts of the detailed queries.

Still very confused...

**Reply by sullynole** [reply]
The total from https://api.weather.gov/alerts/active/count matches the number of alerts from https://api.weather.gov/alerts/active

**Comment by ktrue** [comment]
I've done some more digging and have a couple of questions about the /alerts/active/count return:

1) if a feature entry in /alerts/active lists 4 zones in the feature's UGC list, does that count as 4 alerts (one for each zone listed)?
2) in /alerts/active/count 'zones' entry, there appear a large number of County zones (ssCnnn) entries that are not included in feature entries in the /alerts/active query .. is their presence because a forecast/fire zone (ssZnnn) may intersect one/more County zones?

I've managed to get the counts closer by assuming (1) is correct (count an alert for each zone found in UGC), but...
The total counts land+marine is not equal to total, nor is land+marine+zones ..

```
check-api.php - V1.02 - 13-Oct-2023
 fetching https://api.weather.gov/alerts/active/count
 saved 21051 bytes to alert-counts.json
 JSON decode - No errors

 fetching https://api.weather.gov/alerts/active?status=actual&region_type=land
 saved 372230 bytes to alert-land.json
 JSON decode - No errors

 fetching https://api.weather.gov/alerts/active?status=actual&region_type=marine
 saved 1040754 bytes to alert-marine.json
 JSON decode - No errors

 using current UTC as time
 now UTC for comparisons is Fri, 13 Oct 2023 18:54:23 +0000
 alerts stats will include expired ones.
 alerts/active/count   land=389 v.s. zoneAlerts=456 v.s. count=78 alert features
 alerts/active/count marine=315 v.s. zoneAlerts=448 v.s. count=315 alert features
 alerts/active/count  zones=966 with alerts
 alerts/active/count has 1342 total alerts from 966 zones listed v.s. totZoneAlerts=904

 Analyzing LAND alerts
..Done - 357 seenZones count

 Analyzing MARINE alerts
..Done - 627 seenZones count total

 Checking Zone counts in /alerts/active/count v.s. zones seen in Land and Marine alerts UGC entries.

 These zones had #alerts found in /alerts/active/count but different #alerts counted in seenZones from UGC entries in alerts.
    FLC021:2>1, SDC085:2>1

  All zones seen in /alerts/active/count zone list were found in UGC entries in alerts
   (excepting the 339 county zones not listed in UGC entries)
..Done. 2 mismatches found.

 Checking Zones seen in Land and Marine alerts UGC entries v.s /alerts/active/count zones.

 These zones had #alerts counted in seenZones from UGC entries in alerts but different #alerts found in /alerts/active/count .
    FLC021:1<2, SDC085:1<2

  All zones seen in alerts UGC entries were in /alerts/active/count zone list.
..Done. 2 mismatches found.
```
I've attached the revised check-api.php program V1.02
add ?details=y to see the detail entry decoding.
[check-api-V1.02.zip](https://github.com/weather-gov/api/files/12897700/check-api-V1.02.zip)

**Reply by sullynole** [reply]
1) I am also wondering why doesn't the "land" number from https://api.weather.gov/alerts/active/count match the number of features returned from https://api.weather.gov/alerts/active?region_type=land 

It does seem that it should given the "marine" number matches the number of features here https://api.weather.gov/alerts/active?region_type=marine and the Schema defines this value as "The total number of active alerts affecting land zones"

2) The schema defines the 'zones' counts as "Active alerts by NWS public zone or county code" so it seems that these counts would be a reverse lookup on whether an alert's geometry affects this zone.  These totals may not add up to the total, land, marine values at the top.

![image](https://github.com/weather-gov/api/assets/16312636/688f3d49-139e-414f-8a1d-86aaf2db35f2)
